<a href="https://colab.research.google.com/github/RenjithRT/DSA_Assignments/blob/main/Day_11_activity_students_copy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


### Supervised Learning
### Activity: Building a Student Intervention System

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail or pass. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: This is a Classification supervised learning model**  

**In machine learning, classification refers to a predictive modeling problem where a class label is predicted for a given example of input data.**

**It is called classification problem because classification problems include for our dataset target variable "passed" have yes or no values that is the students passed the course and we have to find out if the students will pass or fail.**

> **This is because there possibly two discrete outcomes, typical of a classification problem.**
* **Students who need early intervention**.
* **Students who do not need early intervention**.

> **We can classify accordingly with a binary outcome such as**:
* **Yes, 1, for students who need early intervention**.
* **No, 0, for students who do not need early intervention**.

**Evidently, we are not trying to predict a continuous outcome, hence this is not a regression problem**.

### Question-2
load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [ ]:
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# Read student data
student_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Week 11/Activity/student-data.csv')

In [ ]:
# Display the data
student_data.head()

school sex  age address famsize  ... Dalc  Walc  health absences passed
0     GP   F   18       U     GT3  ...    1     1       3        6     no
1     GP   F   17       U     GT3  ...    1     1       3        4     no
2     GP   F   15       U     LE3  ...    2     3       3       10    yes
3     GP   F   15       U     GT3  ...    1     1       5        2    yes
4     GP   F   16       U     GT3  ...    1     2       5        4    yes

[5 rows x 31 columns]

In [ ]:
# Display the data in a diagonal view
student_data.head().T

0        1        2         3       4
school           GP       GP       GP        GP      GP
sex               F        F        F         F       F
age              18       17       15        15      16
address           U        U        U         U       U
famsize         GT3      GT3      LE3       GT3     GT3
Pstatus           A        T        T         T       T
Medu              4        1        1         4       3
Fedu              4        1        1         2       3
Mjob        at_home  at_home  at_home    health   other
Fjob        teacher    other    other  services   other
reason       course   course    other      home    home
guardian     mother   father   mother    mother  father
traveltime        2        1        1         1       1
studytime         2        2        2         3       2
failures          0        0        3         0       0
schoolsup       yes       no      yes        no      no
famsup           no      yes       no       yes     yes
paid             no       no      yes       yes     yes
activities       no       no       no       yes      no
nursery         yes       no      yes       yes     yes
higher          yes      yes      yes       yes     yes
internet         no      yes      yes       yes      no
romantic         no       no       no       yes      no
famrel            4        5        4         3       4
freetime          3        3        3         2       3
goout             4        3        2         2       2
Dalc              1        1        2         1       1
Walc              1        1        3         1       2
health            3        3        3         5       5
absences          6        4       10         2       4
passed           no       no      yes       yes     yes

In [ ]:
# Shape of the data
student_data.shape

(395, 31)

In [ ]:
# Summary of the data
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

From the student dataset :
* there are 395 samples(Rows) with 31 features(Columns)
* the last column 'passed' is the target variable in the student dataset (whether the student graduated or didn't graduate)

### Question-3
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [ ]:
# Calculate number of students
n_students = student_data.shape[0]

In [ ]:
# Calculate number of features
n_features = student_data.shape[1]-1

In [ ]:
# Calculate passing students
n_passed = student_data['passed'].value_counts()[0]

In [ ]:
# Calculate failing students
n_failed = student_data['passed'].value_counts()[1]

In [ ]:
# Calculate graduation rate
grad_rate = student_data['passed'].value_counts(normalize=True)*100

In [ ]:
# Print the results
print("Total number of students: {}".format(n_students))
print("The total number of features for each student: {}".format(n_features))
print("The number of those students who passed: {}".format(n_passed))
print('The number of those students who failed: {}'.format(n_failed))
print("Graduation rate of the class: Graduated is {:.2f}%".format(grad_rate[0]),
      "and not graduated is {:.2f}%.".format(grad_rate[1]))

Total number of students: 395
The total number of features for each student: 30
The number of those students who passed: 265
The number of those students who failed: 130
Graduation rate of the class: Graduated is 67.09% and not graduated is 32.91%.


## Preparing the Data
you will prepare the data for modeling, training and testing.

### Question-4 Identify feature and target columns


separate the student data into feature and target columns to see if any features are non-numeric.

In [ ]:
# Extract feature columns

In [ ]:
feature_cols = list(student_data.columns[:-1])

In [ ]:
# Extract target column 'passed'

In [ ]:
target_col = student_data.columns[-1] 

In [ ]:
# Separate the data into feature data and target data (X and y, respectively)

In [ ]:
X = student_data[feature_cols]
y = student_data[target_col]

In [ ]:
# Show the feature information by printing the first five rows
print("\nFeature values:")
print(X.head())


Feature values:
  school sex  age address famsize  ... goout  Dalc  Walc health absences
0     GP   F   18       U     GT3  ...     4     1     1      3        6
1     GP   F   17       U     GT3  ...     3     1     1      3        4
2     GP   F   15       U     LE3  ...     2     2     3      3       10
3     GP   F   15       U     GT3  ...     2     1     1      5        2
4     GP   F   16       U     GT3  ...     2     1     2      5        4

[5 rows x 30 columns]


### Question-5 Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [ ]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X = preprocess_features(X)
print("Processed feature columns ({} total features):\n{}".format(len(X.columns), list(X.columns)))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


In [ ]:
# Display the data in a diagonal view
X.head().T

0   1   2   3   4
school_GP           1   1   1   1   1
school_MS           0   0   0   0   0
sex_F               1   1   1   1   1
sex_M               0   0   0   0   0
age                18  17  15  15  16
address_R           0   0   0   0   0
address_U           1   1   1   1   1
famsize_GT3         1   1   0   1   1
famsize_LE3         0   0   1   0   0
Pstatus_A           1   0   0   0   0
Pstatus_T           0   1   1   1   1
Medu                4   1   1   4   3
Fedu                4   1   1   2   3
Mjob_at_home        1   1   1   0   0
Mjob_health         0   0   0   1   0
Mjob_other          0   0   0   0   1
Mjob_services       0   0   0   0   0
Mjob_teacher        0   0   0   0   0
Fjob_at_home        0   0   0   0   0
Fjob_health         0   0   0   0   0
Fjob_other          0   1   1   0   1
Fjob_services       0   0   0   1   0
Fjob_teacher        1   0   0   0   0
reason_course       1   1   0   0   0
reason_home         0   0   0   1   1
reason_other        0   0   1   0   0
reason_reputation   0   0   0   0   0
guardian_father     0   1   0   0   1
guardian_mother     1   0   1   1   0
guardian_other      0   0   0   0   0
traveltime          2   1   1   1   1
studytime           2   2   2   3   2
failures            0   0   3   0   0
schoolsup           1   0   1   0   0
famsup              0   1   0   1   1
paid                0   0   1   1   1
activities          0   0   0   1   0
nursery             1   0   1   1   1
higher              1   1   1   1   1
internet            0   1   1   1   0
romantic            0   0   0   1   0
famrel              4   5   4   3   4
freetime            3   3   3   2   3
goout               4   3   2   2   2
Dalc                1   1   2   1   1
Walc                1   1   3   1   2
health              3   3   3   5   5
absences            6   4  10   2   4

### Question - 6 Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. you will need to implement the following:
- Randomly shuffle and split the data (`X`, `y`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [ ]:
# splitting the data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=95)

In [ ]:
# Show the results of the split
print("Training set has {} samples.".format(X_train.shape[0]))
print("Testing set has {} samples.".format(X_test.shape[0]))

Training set has 300 samples.
Testing set has 95 samples.


### Question - 7  Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data and measure the accuracy score.

**Answer**

The 3 supervised learning models.

1. **Logistic Regression**
2. **Support Vector Machines**
3. **Decision Tree**

But before we move on to cover the 3 supervised learning models, we will be discussing about the data itself because it is an important discussion to determine if the model makes a good candidate for the problem at hand.

**DATA OVERVIEW**

1. Skewed classes:

As we can see, there is almost twice as many students who passed compared to students who failed.
* Number of students who passed: 265 (majority class)
* Number of students who failed: 130 (minority class)

This would pose problems when we are splitting the data.
* The training test could be populated with mostly the majority class and the testing set could be populated with the minority class. This would affect the accuracy calculated.
* Hence, there should be emphasis on how we split the data and which metric to choose.
> * Splitting the data: Stratified KFold
> * Metrics to choose
>> * Precision and recall
2. Lack of data:

There is a lack of examples in the dataset.
* 395 students

This would have implications on some algorithms that require more data.
* Generally we want more data, except when we are facing a high bias problem.
* In this case, we should keep to simpler algorithms.
3. Too many features:

For such a small dataset with 395 students, we have a staggeringly high number of features.
* 48 features

###  Model Application
*List three supervised learning models that are appropriate for this problem. What are the general applications of each model? What are their strengths and weaknesses? Given what you know about the data, why did you choose these models to be applied?*

In [ ]:
#explaination

The 3 supervised learning models.

1. Logistic Regression
2. Support Vector Machines(SVM)
3. Decision Tree

Let's discuss with one by one for understand the models

## 1. Logistic Regression

In statistics, the logistic model (or logit model) is used to model the probability of a certain class or event existing such as pass/fail, win/lose, alive/dead or healthy/sick. This can be extended to model several classes of events such as determining whether an image contains a cat, dog, lion, etc. Each object being detected in the image would be assigned a probability between 0 and 1, with a sum of one.

For our student student data the target variable 'passed' have yes and no categorical values as it means students passed the course or not.

Strengths
* Many ways to regularize the model to tolerate some errors and avoid over-fitting
* Unlike Naive Bayes, we do not have to worry about correlated features
* Unlike Support Vector Machines, we can easily take in new data using an online gradient descent method

Weaknesses
* Requires observations to be independent of one another
* It aims to predict based on independent variables, if there are not properly identified, Logistic Regression provides little predictive value

Suitability for the problem
* Many features may be correlated
 * Regularization to prevent overfitting due to dataset having many features

## 2. Support Vector Machines (SVMs)
Support vector machines (SVMs) are a set of supervised learning methods used for classification, regression and outliers detection. It is best suited for Classification model.

Strengths
* Effective in high dimensional spaces.

* Still effective in cases where number of dimensions is greater than the number of samples.

* Uses a subset of training points in the decision function (called support vectors), so it is also memory efficient.

* Versatile: different Kernel functions can be specified for the decision function. Common kernels are provided, but it is also possible to specify custom kernels.

Weaknesses
* If the number of features is much greater than the number of samples, avoid over-fitting in choosing Kernel functions and regularization term is crucial.

* SVMs do not directly provide probability estimates, these are calculated using an expensive five-fold cross-validation (see Scores and probabilities, below).

Suitability for the problem
* Many features may be correlated
 * Regularization to prevent overfitting due to dataset having many features

## 3. Decision Tree
Decision Trees (DTs) are a non-parametric supervised learning method used for classification and regression. The goal is to create a model that predicts the value of a target variable by learning simple decision rules inferred from the data features. A tree can be seen as a piecewise constant approximation.

Strengths
* Simple to understand and to interpret. Trees can be visualised.
* Requires little data preparation. Other techniques often require data normalisation, dummy variables need to be created and blank values to be removed. Note however that this module does not support missing values.
* The cost of using the tree (i.e., predicting data) is logarithmic in the number of data points used to train the tree.
* Able to handle both numerical and categorical data. However scikit-learn implementation does not support categorical variables for now. Other techniques are usually specialised in analysing datasets that have only one type of variable. See algorithms for more information.
* Able to handle multi-output problems.
* Uses a white box model. If a given situation is observable in a model, the explanation for the condition is easily explained by boolean logic. By contrast, in a black box model (e.g., in an artificial neural network), results may be more difficult to interpret.
* Possible to validate a model using statistical tests. That makes it possible to account for the reliability of the model.
* Performs well even if its assumptions are somewhat violated by the true model from which the data were generated.

Weaknesses
* Decision-tree learners can create over-complex trees that do not generalise the data well. This is called overfitting. Mechanisms such as pruning, setting the minimum number of samples required at a leaf node or setting the maximum depth of the tree are necessary to avoid this problem.
* Decision trees can be unstable because small variations in the data might result in a completely different tree being generated. This problem is mitigated by using decision trees within an ensemble.
* Predictions of decision trees are neither smooth nor continuous, but piecewise constant approximations as seen in the above figure. Therefore, they are not good at extrapolation.
* The problem of learning an optimal decision tree is known to be NP-complete under several aspects of optimality and even for simple concepts. Consequently, practical decision-tree learning algorithms are based on heuristic algorithms such as the greedy algorithm where locally optimal decisions are made at each node. Such algorithms cannot guarantee to return the globally optimal decision tree. This can be mitigated by training multiple trees in an ensemble learner, where the features and samples are randomly sampled with replacement.
* There are concepts that are hard to learn because decision trees do not express them easily, such as XOR, parity or multiplexer problems.
* Decision tree learners create biased trees if some classes dominate. It is therefore recommended to balance the dataset prior to fitting with the decision tree.

Suitability for the problem
* Allow us to analyze fully the possible consequences of a decision.
* Provide a framework to quantify the values of outcomes and the probabilities of achieving them.
* Help us to make the best decisions on the basis of existing information and best guesses.

In [ ]:
# Import the three supervised learning models from sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

In [ ]:
# fit model-1  on traning data 

In [ ]:
logistic_model = LogisticRegression()
logistic_model.fit(X_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


LogisticRegression()

In [ ]:
# predict on the test data 

In [ ]:
y_predict = logistic_model.predict(X_test)

In [ ]:
# calculate the accuracy score

In [ ]:
from sklearn.metrics import accuracy_score
print("Accuracy score is:",accuracy_score(y_test,y_predict))

Accuracy score is: 0.7157894736842105


In [ ]:
# fit the model-2 on traning data and predict on the test data and measure the accuracy

In [ ]:
# linear classifier
svm_linear = SVC(kernel='linear')
svm_linear.fit(X_train, y_train)
svm_linear_ypred = svm_linear.predict(X_test)
print("The accuracy score using linear svm is ", accuracy_score(y_test,svm_linear_ypred))

# Polynomial SVM
svm_poly = SVC(kernel ='poly', degree=10)
svm_poly.fit(X_train, y_train)
svm_poly_ypred = svm_poly.predict(X_test)
print("The accuracy score using polynomial svm is ", accuracy_score(y_test,svm_poly_ypred))

# Radial Basis Function (RBF) SVM
svm_rbf = SVC(kernel ='rbf')
svm_rbf.fit(X_train, y_train)
svm_rbf_ypred = svm_rbf.predict(X_test)
print("The accuracy score using Radial Basis Function svm is ", accuracy_score(y_test,svm_rbf_ypred))

The accuracy score using linear svm is  0.6947368421052632
The accuracy score using polynomial svm is  0.6631578947368421
The accuracy score using Radial Basis Function svm is  0.631578947368421


In [ ]:
# fit the model-3 on traning data and predict on the test data and measure the accuracy

In [ ]:
dt_model = DecisionTreeClassifier()
dt_model.fit(X_train, y_train)
dt_pred = dt_model.predict(X_test)
print("The accuracy score using Decision Tree classifier is ", accuracy_score(y_test,dt_pred))

The accuracy score using Decision Tree classifier is  0.5263157894736842


# Final Verdict

**We used Logistic Regression, Support Vector Machines(SVMs) and Decision tree models for predict whether the students will pass or fail. From these models, Logistic Regression model performed well with an accuracy of 71.57% comparing with SVMs and Decision Tree algorithms.** 

**From the above analysis, we received best result in Logistic Regression model and hence we conclude that this model is far better than the Support Vector Machines(SVMs) and Decision tree for student data analysis. This is not 100% good model for student data analysis. We have other good models that we can use for getting good accuracy and F1 score to improve the model performance. Also if we scale the data, the model's accuracy will improve and the performance increases.**